File: joiner-core/src/test/java/cz/encircled/joiner/test/config/hint/HintQueryFeature.java
package cz.encircled.joiner.test.config.hint;

import com.google.common.collect.Multimap;
import com.querydsl.jpa.impl.AbstractJPAQuery;
import com.querydsl.jpa.impl.JPAQuery;
import cz.encircled.joiner.query.JoinerQuery;
import cz.encircled.joiner.query.QueryFeature;
import cz.encircled.joiner.test.core.TestException;
import org.springframework.util.ReflectionUtils;

import java.lang.reflect.Field;
import java.util.Collection;

/**
 * @author Kisel on 04.02.2016.
 */
public class HintQueryFeature implements QueryFeature {

    @Override
    public <T, R> JoinerQuery<T, R> before(final JoinerQuery<T, R> request) {
        return request;
    }

    @Override
    public <T, R> JPAQuery<R> after(JoinerQuery<T, R> request, JPAQuery<R> query) {
        Field f = ReflectionUtils.findField(AbstractJPAQuery.class, "hints");
        ReflectionUtils.makeAccessible(f);
        Multimap<String, Object> field = (Multimap<String, Object>) ReflectionUtils.getField(f, query);

        Object value = ((Collection) field.get("testHint")).iterator().next();
        if (!"testHintValue".equals(value)) {
            throw new TestException();
        }

        return query;
    }
}

File: joiner-core/src/test/java/cz/encircled/joiner/test/config/hint/HintQueryFeature.kt
package cz.encircled.joiner.test.config.hint

import com.google.common.collect.Multimap
import com.mysema.query.jpa.impl.AbstractJPAQuery
import com.mysema.query.jpa.impl.JPAQuery
import cz.encircled.joiner.query.Q
import cz.encircled.joiner.query.QueryFeature
import cz.encircled.joiner.test.core.TestException
import org.springframework.util.ReflectionUtils

/**
 * @author Kisel on 04.02.2016.
 */
class HintQueryFeature : QueryFeature {

    override fun <T> before(request: Q<T>): Q<T> {
        return request
    }

    override fun after(request: Q<*>, query: JPAQuery): JPAQuery {
        val f = ReflectionUtils.findField(AbstractJPAQuery<*>::class.java!!, "hints")
        ReflectionUtils.makeAccessible(f)
        val field = ReflectionUtils.getField(f, query) as Multimap<String, Any>

        val value = (field.get("testHint") as Collection<*>).iterator().next()
        if ("testHintValue" != value) {
            throw TestException()
        }

        return query
    }

}
